lib/pbench/server/api/resources/users_api.py

```diff
@@ -231,24 +233,38 @@ def post(self):
 
         # Add the new auth token to the database for later access
         try:
-            token = ActiveTokens(token=auth_token)
-            # TODO: Decide on the auth token limit per user
-            user.update(auth_tokens=token)
-
-            self.logger.info("New auth token registered for user {}", user.username)
+            token = ActiveToken(auth_token, expiration)
+            user.add_token(token)
         except IntegrityError:
-            self.logger.warning(
+            current_app.logger.warning(
                 "Duplicate auth token got created, user might have tried to re-login immediately"
             )
             abort(HTTPStatus.CONFLICT, message="Login collision; please wait and retry")
         except SQLAlchemyError as e:
-            self.logger.error(
+            current_app.logger.error(
                 "SQLAlchemy Exception while logging in a user {}", type(e)
             )
             abort(HTTPStatus.INTERNAL_SERVER_ERROR, message="INTERNAL ERROR")
         except Exception:
-            self.logger.exception("Exception while logging in a user")
+            current_app.logger.exception("Exception while logging in a user")
             abort(HTTPStatus.INTERNAL_SERVER_ERROR, message="INTERNAL ERROR")
+        else:
+            current_app.logger.info(
+                "New auth token registered for user {}", user.username
+            )
+
+        # Remove any expired tokens for this user
+        try:
+            Database.db_session.query(ActiveToken).filter_by(user_id=user.id).filter(
+                ActiveToken.expiration <= datetime.now(timezone.utc)
+            ).delete()
+            Database.db_session.commit()
+        except Exception:
+            Database.db_session.rollback()
+            current_app.logger.exception(
+                "Error encountered while removing expired tokens for user {}",
+                user.username,
+            )
```

Comment:
> For some reason, having this code here feels weird to me -- removing expired tokens doesn't really have anything to do with logging in -- so this is sort of opportunistic housekeeping (e.g., it's not a high-volume or particularly performance-sensitive path, so we'll just do this thing here so that we know that it will happen occasionally...instead of deferring it to a proper housekeeping task).

Right.  Prior to the change and after this change we don't have a "proper" house-keeping task because these "Active Tokens" or "Auth Tokens" or whatever are "going away".

It was the simplest place to put it off of the very heavily traffic token check path.

> If we're going to keep it here, we might as well drop the `filter_by(user_id=user.id)`: an expired token is an expired token, and there's no reason to only delete some of them....

Sure.